cmd/print.go

```diff
@@ -43,55 +46,139 @@ var (
+		os.Exit(1)
+	}
+
+	pr.outputResult(httpRoutes, gateways)
+
+	return nil
+}
+
+func (pr *PrintRunner) outputResult(httpRoutes []gatewayv1beta1.HTTPRoute, gateways []gatewayv1beta1.Gateway) {
+	for i := range gateways {
+		err := pr.resourcePrinter.PrintObj(&gateways[i], os.Stdout)
 		if err != nil {
-			return err
+			fmt.Printf("# Error printing %s HTTPRoute: %v\n", gateways[i].Name, err)
 		}
-		namespaceFilter, err := getNamespaceFilter(namespace, allNamespaces)
+	}
+
+	for i := range httpRoutes {
+		err := pr.resourcePrinter.PrintObj(&httpRoutes[i], os.Stdout)
 		if err != nil {
-			return err
+			fmt.Printf("# Error printing %s HTTPRoute: %v\n", httpRoutes[i].Name, err)
 		}
-		i2gw.Run(resourcePrinter, namespaceFilter, inputFile)
-		return nil
-	},
+	}
 }
 
-// getResourcePrinter returns a specific type of printers.ResourcePrinter
-// based on the provided outputFormat.
-func getResourcePrinter(outputFormat string) (printers.ResourcePrinter, error) {
-	switch outputFormat {
+// initializeResourcePrinter assign a specific type of printers.ResourcePrinter
+// based on the outputFormat of the printRunner struct.
+func (pr *PrintRunner) initializeResourcePrinter() error {
+	switch pr.outputFormat {
 	case "yaml", "":
-		return &printers.YAMLPrinter{}, nil
+		pr.resourcePrinter = &printers.YAMLPrinter{}
+		return nil
 	case "json":
-		return &printers.JSONPrinter{}, nil
+		pr.resourcePrinter = &printers.JSONPrinter{}
+		return nil
 	default:
-		return nil, fmt.Errorf("%s is not a supported output format", outputFormat)
+		return fmt.Errorf("%s is not a supported output format", pr.outputFormat)
 	}
+
 }
 
-// getNamespaceFilter returns a namespace filter, taking into consideration whether a specific
+// initializeNamespaceFilter generate the corret namespace filter, taking into consideration whether a specific
 // namespace is requested, or all of them are.
-func getNamespaceFilter(requestedNamespace string, useAllNamespaces bool) (string, error) {
+func (pr *PrintRunner) initializeNamespaceFilter() error {
+	// When we should use all namespaces, empty string is used as the filter.
+	if pr.allNamespaces {
+		pr.namespaceFilter = ""
+		return nil
+	}
 
-	// When we should use all namespaces, return an empty string.
-	// This is the first condition since it should override the requestedNamespace,
-	// if specified.
-	if useAllNamespaces {
-		return "", nil
+	if pr.namespace == "" {
+		ns, err := getNamespaceInCurrentContext()
+		if err != nil && pr.inputFile == "" {
+			// When asked to read from the cluster, but getting the current namespace
+			// failed for whatever reason - do not process the request.
+			return err
+		}
+		// If err is nil we got the right filtered namespace.
+		// If the input file is specified, and we failed to get the namespace, use all namespaces.
+		pr.namespaceFilter = ns
+		return nil
 	}
 
-	if requestedNamespace == "" {
-		return getNamespaceInCurrentContext()
+	pr.namespaceFilter = pr.namespace
+	return nil
+}
```

Comment:
I think it is better to have it in a struct for few reasons but here are the main ones; 

1. We initialize a printRunner object and hold most of the necessary things to do the print work in it. 
2. We do not pass (and copy, cause we passed by value) function params, instead we have methods on the object
3. For every new flag/feature we wont not need to change the functions signature

Those are the main things I have in mind. 
@levikobi I am curious to why do you think the testability would be more challenging?

Reply:
Well, the issue I had with it is that given this code block:
```go
	err = pr.initializeNamespaceFilter()
	if err != nil {
		return fmt.Errorf("Failed to initialize namespace filter: %v", err)
	}
```
You can't tell if (and which) variables initializeNamespaceFilter needs, and what it returns. I know it says initialize namespaceFilter, but then you have to go and look if there is actually a namespaceFilter variable.
As opposed to:
```go
	pr.namespaceFilter, err = getNamespaceFilter(namespace, allNamespaces)
	if err != nil {
		return fmt.Errorf("Failed to initialize namespace filter: %v", err)
	}
```
That's why, in my opinion, we are losing readability here.

@LiorLieberman regarding what I said about testability, I think it's just a bit harder to test because it's a bit less readable.

But anyways I see that both of you are in favor of this implementation so I'm good with it too 👍